### PR TITLE
feat: use blockfrost to set the validity range

### PR DESCRIPTION
**Pull request by sk-kitsune**
Resolves https://app.asana.com/0/1203842063837585/1204052052893517

Tested backing and unbacking flows

https://user-images.githubusercontent.com/117076162/224239024-1cf8a59d-8330-4a11-848a-a825c56fa25f.mp4

### CODE REVIEW

**Comment by sk-shishi** [comment]
Please wait until https://github.com/teiki-network/teiki-protocol/pull/152 is merged.

Then we won't have any `TX_TIME_..._PADDING` settings anymore, just `DEFAULT_TX_TTL`.

File: modules/protocol/utils.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import { UnixTimestamp } from "@/modules/business-types";
+import { writeErrorToConsole } from "@/modules/displayable-error";
+
+export async function getTxTimeStart(): Promise<UnixTimestamp> {
+  try {
+    const response = await httpGetLatestBlock();
+    return response.time * 1000;
+  } catch (error) {
+    writeErrorToConsole(error);
+    return TX_TIME_START_PADDING;
```
Comment (sk-shishi) [inline modules/protocol/utils.ts:13]: This looks incorrect.

Comment (sk-kitsune) [reply modules/protocol/utils.ts:13]: Can you elaborate?

Comment (sk-shishi) [reply modules/protocol/utils.ts:13]: The 1st case returns an `UnixTimestamp`, while the 2nd case returns a "time difference / duration".

Comment (sk-kitsune) [reply modules/protocol/utils.ts:13]: oops, my bad

**Comment by sk-shishi** [comment]
@sk-kitsune The protocol PR is merged, commit hash `f6f3055276f3680e0330d7fe371054270551fc9e` (https://github.com/teiki-network/teiki-protocol/commit/f6f3055276f3680e0330d7fe371054270551fc9e)

File: modules/protocol/utils.ts (hunk 1 above)
Comment (sk-kitsune) [reply modules/protocol/utils.ts]: fixed

**Comment by sk-kitsune** [comment]
> @sk-kitsune The protocol PR is merged, commit hash `f6f3055276f3680e0330d7fe371054270551fc9e` ([teiki-network/teiki-protocol@f6f3055](https://github.com/teiki-network/teiki-protocol/commit/f6f3055276f3680e0330d7fe371054270551fc9e))

ok, I will upgrade teiki-protocol to this version

File: modules/protocol/utils.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { httpGetLatestBlock } from "./api/httpGetLatestBlock";
+
+import { UnixTimestamp } from "@/modules/business-types";
+import { writeErrorToConsole } from "@/modules/displayable-error";
+
+const FALLBACK_PADDING = 60000;
+
+export async function getTxTimeStart(): Promise<UnixTimestamp> {
```
Comment (sk-kitsune) [inline modules/protocol/utils.ts:8]: @sk-shishi do you have a more meaningful name for this function?

File: modules/next-backend/logic/getProjectTeikiRewardsByBacker.ts [hunk 3]
```diff
@@ -52,10 +51,9 @@ export async function getProjectTeikiRewardsByBacker(
     ProtocolParamsDatum
   );
 
-  const now = Date.now();
-  const timeProvider = () => now;
+  const txTime = await getTxTimeStart();
 
-  const unstakedAt = getTime({ timeProvider }) - TX_TIME_START_PADDING;
+  const unstakedAt = txTime;
```
Comment (sk-kitsune) [inline modules/next-backend/logic/getProjectTeikiRewardsByBacker.ts:56]: @sk-shishi can you double check this line?

Comment (sk-shishi) [reply modules/next-backend/logic/getProjectTeikiRewardsByBacker.ts:56]: It's correct.

File: modules/protocol/utils.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  try {
+    const response = await httpGetLatestBlock();
+    return response.time * 1000;
+  } catch (error) {
+    writeErrorToConsole(error);
+    const now = Math.floor(Date.now() / 1000) * 1000;
```
Comment (sk-shishi) [inline modules/protocol/utils.ts:14]: There's a trimToSlot function exposed by protocol

File: modules/protocol/utils.ts (hunk 2 above)
Comment (sk-shishi) [reply modules/protocol/utils.ts:8]: Nope, maybe getReferenceTxTime. But I think leaving some comments explaining that nodes use "last block time" to determine the valid time range.

**Review by sk-shishi — APPROVED** [review]

**Pull request merged**